README.md
# AWS Cloud (un)Map

External controller to remove terminated EC2 instances from AWS Cloud Map service.


## How it works

This application scans - at a regular interval - the instances registered to a Cloud Map service and match them with the EC2 instances running in 1+ region: it will then deregister any instance registered in the service which doesn't match a running EC2 instance.

Requisites:
- The instance must be registered in the Cloud Map service with Cloud Map instance id equal to the EC2 instance id
- The instance must be registered in the Cloud Map service with `AWS_INSTANCE_IPV4` attribute (can be the private or public IP address)

How the matching is done:
- A registered instance is considered valid if **both** the instance id and the `AWS_INSTANCE_IPV4` address match a running EC2 instance
- A registered instance is skipped (left untouched) if registered without `AWS_INSTANCE_IPV4` attribute

Safety countermeasures:
- The application logs a warning and do **not** deregister the unmatching instances, in case that would leave the service without registered instance
- The application handles graceful shutdown on `SIGINT` and `SIGTERM`. If such signals are received during a reconciling, it would complete the on-going reconcile before exiting


## How to run it

You have two options to run it:

1. Manually install and run the [`aws-cloud-unmap` Python package](https://pypi.org/project/aws-cloud-unmap/)
   ```
   pip3 install aws-cloud-unmap

   aws-cloud-unmap --service-id srv-12345 --service-region us-east-1 --instances-region us-east-1
   ```

2. Use the [Docker image available on Docker hub](https://hub.docker.com/r/spreaker/aws-cloud-unmap/)
   ```
   docker run --env AWS_ACCESS_KEY_ID="id" --env AWS_SECRET_ACCESS_KEY="secret" spreaker/aws-cloud-unmap --service-id srv-12345 --service-region us-east-1 --instances-region us-east-1
   ```


The cli supports the following arguments:

| Argument                                 | Required | Description |
| ---------------------------------------- | -------- | ----------- |
| `--service-id ID`                        | yes      | AWS CloudMap service ID |
| `--service-region REGION`                | yes      | AWS CloudMap service region |
| `--instances-region REGION [REGION ...]` | yes      | AWS regions where EC2 instances should be checked |
| `--frequency N`                          |          | How frequently the service should be reconciled (in seconds). Defaults to `300` sec |
| `--single-run`                           |          | Run a single reconcile and then exit |
| `--enable-prometheus`                    |          | Enable the Prometheus exporter. Disabled by default |
| `--prometheus-host`                      |          | The host at which the Prometheus exporter should listen to. Defaults to `127.0.0.1` |
| `--prometheus-port`                      |          | The port at which the Prometheus exporter should listen to. Defaults to `9100` |
| `--log-level LOG_LEVEL`                  |          | Minimum log level. Accepted values are: `DEBUG`, `INFO`, `WARNING`, `ERROR`, `CRITICAL`. Defaults to `INFO` |


## Exported metrics

The application features an integrated Prometheus exporter. The following metrics are exported:

| Metric name                                                | Labels       | Description |
| ---------------------------------------------------------- | ------------ | ----------- |
| `aws_cloud_unmap_up`                                       | `service_id` | Always `1`: can be used to check if it's running |
| `aws_cloud_unmap_last_reconcile_success_timestamp_seconds` | `service_id` | The timestamp (in seconds) of the last successful reconciliation |


## Required IAM privileges

In order to successfully run, this application requires the following IAM privileges:

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid":      "DescribeEC2Instances",
      "Effect":   "Allow",
      "Action":   [ "ec2:DescribeInstances" ],
      "Resource": "*"
    },{
      "Sid":      "ListAndDeregisterServiceInstances",
      "Effect":   "Allow",
      "Action":   [
        "servicediscovery:ListInstances",
        "servicediscovery:DeregisterInstance",
        "route53:GetHealthCheck",
        "route53:DeleteHealthCheck",
        "route53:UpdateHealthCheck"
      ],
      "Resource": "*"
    },{
      "Sid":      "UpdateDnsWhileDeregisteringServiceInstances",
      "Effect":   "Allow",
      "Action":   [ "route53:ChangeResourceRecordSets" ],
      "Resource": [
        "ARN-OF-YOUR-ROUTE53-HOSTEDZONE"
      ]
    }
  ]
}
```


## Development

Run the development environment:

```
docker-compose build dev && docker-compose run --rm dev
```

Run tests in the dev environment:

```
python3 -m unittest
```


## License

This software is released under the [MIT license](LICENSE.txt).



CHANGELOG.md
# Changelog

### 2.0.0 (2023-10-05)
- Upgrade to Python 3.11
- Upgrade dependencies

### 1.1.1 (2019-04-04)
- [BUGFIX] Do not fail the reconciling if there are no instances registered to the service - PR [#8](https://github.com/spreaker/aws-cloud-unmap/pull/8)
- Documented required IAM privileges - PR [#7](https://github.com/spreaker/aws-cloud-unmap/pull/7)

### 1.1.0 (2019-04-02)
- Introduced a built-in Prometheus exporter - PR [#6](https://github.com/spreaker/aws-cloud-unmap/pull/6)

### 1.0.4 (2019-03-29)
- [BUGFIX] Fixed memory leak - PR [#5](https://github.com/spreaker/aws-cloud-unmap/pull/5)

### 1.0.3 (2019-03-29)
- First release


tests/mocks.py
from unittest.mock import MagicMock


def mockBotoClient(clients):
    mock = MagicMock()
    mock.side_effect = lambda name, **kwargs: clients[name]

    return mock


def mockServiceInstance(instanceId, ipv4=None):
    instance = {"Id": instanceId, "Attributes": {"AWS_INSTANCE_PORT": "80"}}

    if ipv4:
        instance["Attributes"]["AWS_INSTANCE_IPV4"] = ipv4

    return instance


def mockEC2Instance(instanceId, privateIp=None, publicIp=None, state="running"):
    instance = {"InstanceId": instanceId, "State": {"Name": state}}

    if privateIp:
        instance["PrivateIpAddress"] = privateIp

    if publicIp:
        instance["PublicIpAddress"] = publicIp

    return instance


tests/__init__.py
import logging

# Disable logging when running tests
logging.disable(logging.CRITICAL)


tests/test_aws.py
import unittest
import boto3
from botocore.stub import Stubber
from cloudunmap.aws import listEC2InstancesById, listServiceInstances
from .mocks import mockEC2Instance, mockServiceInstance


class TestAws(unittest.TestCase):
    #
    # listEC2InstancesById()
    #

    def testListEC2InstancesByIdShouldReturnEc2InstancesFilteredByInputIds(self):
        ec2Client = boto3.client("ec2")

        # Mock EC2 client
        stubber = Stubber(ec2Client)
        stubber.add_response(
            "describe_instances",
            {"Reservations": [{"Instances": [mockEC2Instance("i-1", privateIp="172.0.0.1"), mockEC2Instance("i-2", privateIp="172.0.0.2", publicIp="2.2.2.2")]}]},
            {"Filters": [{"Name": "instance-id", "Values": ["i-1", "i-2"]}], "MaxResults": 1000})
        stubber.activate()

        instances = listEC2InstancesById(["i-1", "i-2"], ec2Client)
        self.assertEqual(len(instances), 2)
        self.assertEqual(instances[0]["InstanceId"], "i-1")
        self.assertEqual(instances[1]["InstanceId"], "i-2")

        stubber.assert_no_pending_responses()

    def testListEC2InstancesByIdShouldReturnEmptyListOnNoMatchingInstances(self):
        ec2Client = boto3.client("ec2")

        # Mock EC2 client
        stubber = Stubber(ec2Client)
        stubber.add_response(
            "describe_instances",
            {"Reservations": []},
            {"Filters": [{"Name": "instance-id", "Values": ["i-1", "i-2"]}], "MaxResults": 1000})
        stubber.activate()

        instances = listEC2InstancesById(["i-1", "i-2"], ec2Client)
        self.assertEqual(len(instances), 0)

        stubber.assert_no_pending_responses()

    def testListEC2InstancesByIdShouldRaiseExceptionOnError(self):
        ec2Client = boto3.client("ec2")

        # Mock EC2 client
        stubber = Stubber(ec2Client)
        stubber.add_client_error("describe_instances")
        stubber.activate()

        with self.assertRaises(Exception):
            listEC2InstancesById(["i-1", "i-2"], ec2Client)

        stubber.assert_no_pending_responses()

    #
    # listServiceInstances()
    #

    def testListServiceInstancesShouldReturnInstancesRegisteredToCloudMapService(self):
        sdClient = boto3.client("servicediscovery")

        # Mock Cloud Map client
        stubber = Stubber(sdClient)
        stubber.add_response(
            "list_instances",
            {"Instances": [mockServiceInstance("i-1", ipv4="172.0.0.1"), mockServiceInstance("i-2", ipv4="2.2.2.2")]},
            {"ServiceId": "srv-1", "MaxResults": 100})
        stubber.activate()

        instances = listServiceInstances("srv-1", sdClient)
        self.assertEqual(len(instances), 2)
        self.assertEqual(instances[0]["Id"], "i-1")
        self.assertEqual(instances[1]["Id"], "i-2")

        stubber.assert_no_pending_responses()

    def testListServiceInstancesShouldRaiseExceptionOnError(self):
        sdClient = boto3.client("servicediscovery")

        # Mock Cloud Map client
        stubber = Stubber(sdClient)
        stubber.add_client_error("list_instances")
        stubber.activate()

        with self.assertRaises(Exception):
            listServiceInstances("srv-1", sdClient)

        stubber.assert_no_pending_responses()


cloudunmap/aws.py
from typing import List


def listEC2InstancesById(instanceIds: List[str], ec2Client):
    instances = []

    # Create a paginator
    paginator = ec2Client.get_paginator("describe_instances")

    # Filter instances by instance id, without breaking if
    # an instance ID is missing or invalid
    filters = [{"Name": "instance-id", "Values": instanceIds}]

    # Pick instances from all pages
    for page in paginator.paginate(Filters=filters, PaginationConfig={"PageSize": 1000}):
        if "Reservations" not in page:
            continue

        for reservation in page["Reservations"]:
            if "Instances" in reservation:
                instances.extend(reservation["Instances"])

    return instances


def listServiceInstances(serviceId: str, sdClient):
    instances = []

    # Create a paginator
    paginator = sdClient.get_paginator("list_instances")

    # Pick instances from all pages
    for page in paginator.paginate(ServiceId=serviceId, PaginationConfig={"PageSize": 100}):
        instances.extend(page["Instances"])

    return instances


cloudunmap/unmap.py
import boto3
import botocore
import logging
from typing import List
from .aws import listServiceInstances, listEC2InstancesById


def matchServiceInstanceInRunningInstances(serviceInstance, runningInstances):
    serviceInstanceId = serviceInstance["Id"]
    serviceInstanceIp = serviceInstance["Attributes"]["AWS_INSTANCE_IPV4"]

    for runningInstance in runningInstances:
        # The instance ID must match
        if runningInstance["InstanceId"] != serviceInstanceId:
            continue

        # The service IP must match any of the instance IPs
        if "PublicIpAddress" in runningInstance and runningInstance["PublicIpAddress"] == serviceInstanceIp:
            return True
        elif "PrivateIpAddress" in runningInstance and runningInstance["PrivateIpAddress"] == serviceInstanceIp:
            return True

    return False


def unmapTerminatedInstancesFromService(serviceId: str, serviceRegion: str, instancesRegions: List[str]) -> bool:
    logger = logging.getLogger()
    logger.info(f"Checking EC2 instances registered to service {serviceId} in {serviceRegion}")

    # Customize the boto client config
    botoConfig = botocore.client.Config(connect_timeout=5, read_timeout=15, retries={"max_attempts": 2})

    # Instance Cloud Map client
    sdClient = boto3.client("servicediscovery", config=botoConfig, region_name=serviceRegion)

    # Instance EC2 clients
    ec2Clients = [boto3.client("ec2", config=botoConfig, region_name=region) for region in instancesRegions]

    # List registered instances on CloudMap
    serviceInstances = listServiceInstances(serviceId, sdClient)

    # Skip in case there are no registered instances
    if not serviceInstances:
        logger.info(f"No registered instances in the service {serviceId}. Skipping")
        return True

    # Filter out service instances without the AWS_INSTANCE_IPV4 attribute
    # and extract instance ids
    serviceInstances = list(filter(lambda i: "AWS_INSTANCE_IPV4" in i["Attributes"], serviceInstances))
    serviceInstancesId = list(map(lambda i: i["Id"], serviceInstances))

    # List EC2 instances in all expected regions
    runningInstances = []

    for ec2Client in ec2Clients:
        instances = listEC2InstancesById(serviceInstancesId, ec2Client)
        # Filter out terminated instances
        instances = list(filter(lambda i: i["State"]["Name"] != "shutting-down" and i["State"]["Name"] != "terminated", instances))

        runningInstances += instances

    # Find the list of unmatching instances. The match is done both by
    # instance ID and IP, to avoid edge cases with recycled IPs on different
    # instance IDs
    unmatchingInstances = list(filter(lambda i: not matchServiceInstanceInRunningInstances(i, runningInstances), serviceInstances))

    # Circuit breaker: ensure that we're not going to remove ALL instances
    # from the service
    if len(unmatchingInstances) >= len(serviceInstances):
        logger.warning(f"All instances registered to service {serviceId} appear to not match any running EC2 instance in {instancesRegions}, but skipping deregistering as safe protection")
        return False

    # Remove all unmatching instances from the service
    logger.info(f"Found {len(unmatchingInstances)} instances in service {serviceId} not matching any running EC2 instance in {instancesRegions}")

    for unmatchingInstance in unmatchingInstances:
        logger.warning(f"Deregistering instance {unmatchingInstance['Id']} from service {serviceId} because not matching any running EC2 instance in {instancesRegions}")
        sdClient.deregister_instance(ServiceId=serviceId, InstanceId=unmatchingInstance["Id"])

    logger.info(f"Checked EC2 instances registered to service {serviceId} in {serviceRegion}")
    return True


cloudunmap/cli.py
import argparse
import logging
import time
import sys
import signal
from typing import List
from pythonjsonlogger import jsonlogger
from .unmap import unmapTerminatedInstancesFromService
from prometheus_client import start_http_server, Gauge


# Prometheus metrics
upMetric = Gauge(
    "aws_cloud_unmap_up",
    "Always 1 - can by used to check if it's running",
    labelnames=["service_id"])

lastReconcileTimestampMetric = Gauge(
    "aws_cloud_unmap_last_reconcile_success_timestamp_seconds",
    "The timestamp (in seconds) of the last successful reconciliation",
    labelnames=["service_id"])


def parseArguments(argv: List[str]):
    # Parse arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("--service-id", metavar="ID", required=True, help="AWS CloudMap service ID")
    parser.add_argument("--service-region", metavar="REGION", required=True, help="AWS CloudMap service region")
    parser.add_argument("--instances-region", metavar="REGION", required=True, nargs='+', help="AWS region where EC2 instances should be checked")
    parser.add_argument("--frequency", metavar="N", required=False, type=int, default=300, help="How frequently the service should be reconciled (in seconds)")
    parser.add_argument("--single-run", required=False, default=False, action="store_true", help="Run a single reconcile and then exit")
    parser.add_argument("--enable-prometheus", required=False, default=False, action="store_true", help="Enable the Prometheus exporter")
    parser.add_argument("--prometheus-host", required=False, default="127.0.0.1", help="The host at which the Prometheus exporter should listen to")
    parser.add_argument("--prometheus-port", required=False, default="9100", type=int, help="The port at which the Prometheus exporter should listen to")
    parser.add_argument("--log-level", help="Minimum log level. Accepted values are: DEBUG, INFO, WARNING, ERROR, CRITICAL", default="INFO")

    return parser.parse_args(argv)


def reconcile(serviceId: str, serviceRegion: str, instancesRegion: List[str]):
    logger = logging.getLogger()

    try:
        success = unmapTerminatedInstancesFromService(serviceId, serviceRegion, instancesRegion)
    except Exception as error:
        logger.error(f"An error occurred while reconciling service {serviceId}: {str(error)}")
        success = False

    if success:
        lastReconcileTimestampMetric.labels(serviceId).set(int(time.time()))


def main(args):
    shutdown = False

    # Init logger
    logHandler = logging.StreamHandler()
    formatter = jsonlogger.JsonFormatter("%(asctime)s %(levelname)s %(message)s", datefmt="%Y-%m-%d %H:%M:%S")
    logHandler.setFormatter(formatter)

    logger = logging.getLogger()
    logger.addHandler(logHandler)
    logger.setLevel(args.log_level)

    # Register signal handler
    def _on_sigterm(signal, frame):
        logger.info("Shutting down")
        nonlocal shutdown
        shutdown = True

    signal.signal(signal.SIGINT, _on_sigterm)
    signal.signal(signal.SIGTERM, _on_sigterm)

    # Start Prometheus exporter
    if args.enable_prometheus:
        start_http_server(args.prometheus_port, args.prometheus_host)
        logger.info("Prometheus exporter listening on {host}:{port}".format(port=args.prometheus_port, host=args.prometheus_host))

    # Set the up metric value, which will be steady to 1 for the entire app lifecycle
    upMetric.labels(args.service_id).set(1)

    # Reconcile
    if args.single_run:
        reconcile(args.service_id, args.service_region, args.instances_region)
    else:
        while not shutdown:
            startTime = time.monotonic()
            reconcile(args.service_id, args.service_region, args.instances_region)

            # Honor frequency
            while not shutdown:
                elapsedTime = time.monotonic() - startTime

                if elapsedTime < args.frequency:
                    time.sleep(min(1, args.frequency - elapsedTime))
                else:
                    break


def run():
    main(parseArguments(sys.argv[1:]))


if __name__ == '__main__':
    run()
